### PR TITLE
feat: PPFT training

### PR DESCRIPTION
**Pull request by sarahnlewis**

### CODE REVIEW

File: src/bioemu/training/loss.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+import torch
```
Comment (ryotatomioka) [inline src/bioemu/training/loss.py:1]:
```suggestion
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
import torch
```

File: src/bioemu/training/foldedness.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    heavy_pairs = []
+    for i, j in combinations(heavy_atoms, 2):
+        res_position_i = traj.topology.atom(i).residue.index
+        res_position_j = traj.topology.atom(j).residue.index
+
+        # Check if residues are sufficiently far apart in the sequence:
+        if abs(res_position_i - res_position_j) > SEQUENCE_SEPARATION:
+            heavy_pairs.append((i, j))
+
+            # TODO: why do we need this? This extra pair is redundant
```
Comment (ryotatomioka) [inline src/bioemu/training/foldedness.py:192]: @YuuuXie is this necessary?

File: notebooks/ppft_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+    "        seq=reference_info.sequence, cache_embeds_dir=None\n",
+    "    )\n",
+    "    single_embeds = np.load(single_embeds_file)\n",
+    "    pair_embeds = np.load(pair_embeds_file)\n",
+    "    assert pair_embeds.shape[0] == pair_embeds.shape[1] == n\n",
+    "    assert single_embeds.shape[0] == n\n",
+    "    assert len(single_embeds.shape) == 2\n",
+    "    _, _, n_pair_feats = pair_embeds.shape  # [seq_len, seq_len, n_pair_feats]\n",
+    "\n",
+    "    single_embeds, pair_embeds = torch.from_numpy(single_embeds), torch.from_numpy(pair_embeds)\n",
```
Comment (ryotatomioka) [inline notebooks/ppft_example.ipynb:231]: Can we combine these lines with `np.load` to avoid the same variable having different types?

File: notebooks/ppft_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "19d4881e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from bioemu.sample import main as sample_main\n",
```
[the remark this answers is not in the document]
Comment (sarahnlewis) [reply notebooks/ppft_example.ipynb:56]: I tried changing this to use `generate_batch`. It was tidier in some ways, but I did not like that the samples were no longer saved as .xtc and .pdb. I think the .xtc and .pdb are nice for curious users who might want to view the structures, so I propose to leave this using `sample.main`.

**Review by ryotatomioka — APPROVED** [review]
Thank you, LGTM

File: src/bioemu/model_utils.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import os
```
Comment (ryotatomioka) [inline src/bioemu/model_utils.py:1]:
```suggestion
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
import os
```

File: notebooks/rollout.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+_target_: bioemu.training.loss._rollout
```
Comment (ryotatomioka) [inline notebooks/rollout.yaml:1]:
```suggestion
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
_target_: bioemu.training.loss._rollout
```

File: tests/training/test_loss.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+from pathlib import Path
```
Comment (ryotatomioka) [inline tests/training/test_loss.py:1]:
```suggestion
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.
from pathlib import Path
```
